

# Hyperparameter Tuning (GridSearch vs. RandomizedSearch) and Cross-Validation (CV)

In machine learning, model performance is heavily influenced by the choice of hyperparameters—settings that govern the training process and structure of the model. Hyperparameter tuning is the process of systematically searching for the optimal combination of these parameters to enhance model accuracy and generalization.

Cross-validation is a robust technique used to assess the performance of a model by partitioning the dataset into subsets. This method allows for a more reliable evaluation of the model's ability to generalize to unseen data. By training and validating the model on different subsets, cross-validation helps mitigate issues such as overfitting and provides a clearer picture of how the model will perform in real-world scenarios.

Together, hyperparameter tuning and cross-validation form a critical part of the machine learning workflow, ensuring that models are both well-optimized and capable of making accurate predictions on new data.

`GridSearchCV` and `RandomizedSearchCV` are essential techniques for hyperparameter tuning in machine learning. While both methods aim to optimize model performance, they differ significantly in their approach to exploring the hyperparameter space.




## 1. **GridSearchCV**


In [4]:
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestRegressor
from sklearn.datasets import make_regression

# Create a synthetic dataset
X_train, y_train = make_regression(n_samples=100, n_features=10, noise=0.1, random_state=42)


In [6]:

import warnings

# Suppress warnings in a specific block of code
with warnings.catch_warnings():
    warnings.simplefilter("ignore")

# Define the parameter grid for GridSearchCV

param_grid = {
    'n_estimators': [50, 100, 200],
    'max_features': ['auto', 'sqrt', 'log2'],
    'max_depth': [4, 5, 6, 7, 8]
}

# Create a Random Forest Regressor
rf = RandomForestRegressor(random_state=42)

# Create GridSearchCV object
grid_search = GridSearchCV(estimator=rf, param_grid=param_grid, cv=5, n_jobs=-1, verbose=0)

# Fit GridSearchCV
grid_search.fit(X_train, y_train)

# Best parameters and best score
print(f'Best parameters: {grid_search.best_params_}')
print(f'Best cross-validation score: {grid_search.best_score_}')

/opt/anaconda3/lib/python3.11/site-packages/sklearn/ensemble/_forest.py:413: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features=1.0` or remove this parameter as it is also the default value for RandomForestRegressors and ExtraTreesRegressors.
  warn(
/opt/anaconda3/lib/python3.11/site-packages/sklearn/ensemble/_forest.py:413: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features=1.0` or remove this parameter as it is also the default value for RandomForestRegressors and ExtraTreesRegressors.
  warn(
/opt/anaconda3/lib/python3.11/site-packages/sklearn/ensemble/_forest.py:413: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features=1.0` or remove this parameter as it is also the default value for RandomFore

Best parameters: {'max_depth': 8, 'max_features': 'auto', 'n_estimators': 50}
Best cross-validation score: 0.6752852312670674



## Explanation of Code

In this section, we trained a **Random Forest Regressor**, an ensemble learning technique that constructs multiple decision trees during the training process. For regression tasks, it outputs the average prediction from these individual trees. This ensemble approach enhances accuracy and reduces the risk of overfitting.

## GridSearchCV

To optimize hyperparameters, we employed **Grid Search**, which systematically evaluates the model's performance for each combination of hyperparameters defined in the `parameter_grid`:

- **`n_estimators`:** Specifies the number of trees in the random forest.
- **`max_features`:** Sets the maximum number of features to consider when searching for the best split.
- **`max_depth`:** Determines the maximum depth of the trees in the random forest.

### **k-Fold Cross-Validation**

We utilized **5-fold cross-validation** (`cv=5`), which involves dividing the dataset into five subsets, or "folds." The model is trained and validated five times, with each iteration using a different fold as the validation set. In each iteration, one fold is reserved for validation while the remaining four folds are used for training. This approach provides a more accurate estimate of the model's performance on unseen data, as it assesses the model's ability to generalize across various subsets of the dataset.

### Best Parameters

- **`max_depth`: 8**  
  A depth of 8 allows the trees to grow up to eight levels deep, enabling the model to capture more complex patterns in the data while minimizing the risk of overfitting.

- **`max_features`: 'auto'**  
  By setting this parameter to 'auto', the model uses the square root of the total number of features. This approach is commonly used in classification tasks and helps reduce overfitting by introducing randomness into the model.

- **`n_estimators`: 50**  
  A value of 50 indicates that the model will utilize 50 individual decision trees for making predictions. While increasing the number of trees can enhance performance, it also raises computational costs. In this case, 50 trees strike a good balance between performance and efficiency.

### Best Cross-Validation Score

The score of **0.6753** reflects the average performance of the model across the cross-validation folds, indicating that the model explains approximately 67.53% of the variance in the target variable based on the training data. While this is a moderate score, it suggests that the model captures some underlying patterns in the data, though there may still be opportunities for impro


## 2.**RandomizedSearchCV** 


In [7]:
from sklearn.model_selection import RandomizedSearchCV
from sklearn.ensemble import RandomForestRegressor
from scipy.stats import randint

# Define the parameter distribution
param_dist = {
    'n_estimators': randint(50, 200),
    'max_features': ['auto', 'sqrt', 'log2'],
    'max_depth': randint(4, 8),
    'min_samples_split': randint(2, 10),
    'min_samples_leaf': randint(1, 4)
}

# Create a Random Forest Regressor
rf = RandomForestRegressor(random_state=42)

# Create RandomizedSearchCV object
random_search = RandomizedSearchCV(
    estimator=rf,
    param_distributions=param_dist,
    n_iter=100,
    cv=5,
    n_jobs=-1,
    verbose=0,
    random_state=42
)

# Fit RandomizedSearchCV
random_search.fit(X_train, y_train)

# Output best parameters and best score
print(f'Best parameters: {random_search.best_params_}')
print(f'Best cross-validation score: {random_search.best_score_}')


/opt/anaconda3/lib/python3.11/site-packages/sklearn/ensemble/_forest.py:413: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features=1.0` or remove this parameter as it is also the default value for RandomForestRegressors and ExtraTreesRegressors.
  warn(
/opt/anaconda3/lib/python3.11/site-packages/sklearn/ensemble/_forest.py:413: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features=1.0` or remove this parameter as it is also the default value for RandomForestRegressors and ExtraTreesRegressors.
  warn(
/opt/anaconda3/lib/python3.11/site-packages/sklearn/ensemble/_forest.py:413: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features=1.0` or remove this parameter as it is also the default value for RandomFore

Best parameters: {'max_depth': 7, 'max_features': 'auto', 'min_samples_leaf': 2, 'min_samples_split': 3, 'n_estimators': 58}
Best cross-validation score: 0.6730052426299349



## Explanation of Code

In this section, we trained a **Random Forest Regressor**, an ensemble learning technique that constructs multiple decision trees during the training process. For regression tasks, it outputs the average prediction from these individual trees. This ensemble approach enhances accuracy and reduces the risk of overfitting.

## RandomizedSearchCV

To optimize hyperparameters, we employed **Randomized Search**, which samples a fixed number of hyperparameter combinations from the specified distributions defined in the `param_dist`:

- **`n_estimators`:** Specifies the number of trees in the random forest, sampled randomly between 50 and 200.
- **`max_features`:** Sets the maximum number of features to consider when searching for the best split, with options including 'auto', 'sqrt', and 'log2'.
- **`max_depth`:** Determines the maximum depth of the trees in the random forest, sampled randomly between 4 and 8.
- **`min_samples_split`:** Specifies the minimum number of samples required to split an internal node, sampled randomly between 2 and 10.
- **`min_samples_leaf`:** Defines the minimum number of samples required to be at a leaf node, sampled randomly between 1 and 4.

### **k-Fold Cross-Validation**

We utilized **5-fold cross-validation** (`cv=5`), which involves dividing the dataset into five subsets, or "folds." The model is trained and validated five times, with each iteration using a different fold as the validation set. In each iteration, one fold is reserved for validation while the remaining four folds are used for training. This approach provides a more accurate estimate of the model's performance on unseen data, as it assesses the model's ability to generalize across various subsets of the dataset.

### Best Parameters

- **`max_depth`: 7**  
  A depth of 7 allows the trees to grow up to seven levels deep, enabling the model to capture complex patterns in the data while minimizing the risk of overfitting.

- **`max_features`: 'auto'**  
  By setting this parameter to 'auto', the model uses all available features for splitting, which can enhance performance while maintaining a balance between complexity and interpretability.

- **`min_samples_leaf`: 2**  
  A value of 2 indicates that a minimum of two samples is required to be at a leaf node, which helps prevent overfitting by ensuring that leaf nodes contain enough data points.

- **`min_samples_split`: 3**  
  This parameter specifies that at least three samples are needed to split an internal node, promoting more robust splits and reducing the risk of overfitting.

- **`n_estimators`: 58**  
  A value of 58 indicates that the model will utilize 58 individual decision trees for making predictions. This number strikes a balance between performance and computational efficiency.

### Best Cross-Validation Score

The score of **0.6730** reflects the average performance of the model across the cross-validation folds, indicating that the model explains approximately 67.30% of the variance in the target variable based on the training data. While this is a moderate score, it suggests that the model captures some underlying patterns in the data, though there may still be opportunities for improvement.





## Comparison RandomizedSearch vs GridSearch:

- **GridSearchCV** is exhaustive and deterministic, making it thorough but potentially computationally expensive.
- **RandomizedSearchCV** is more efficient and scalable, making it preferable for larger hyperparameter spaces, though it may yield different results across runs unless a random seed is specified.

Choosing between `GridSearchCV` and `RandomizedSearchCV` depends on the size of the hyperparameter space, available computational resources, and the desired balance between thoroughness and efficiency:

| Feature                | GridSearchCV                               | RandomizedSearchCV                          |
|------------------------|-------------------------------------------|--------------------------------------------|
| **Search Method**      | Exhaustive search over specified parameters | Random sampling of parameter combinations   |
| **Computational Cost** | High, especially with large parameter spaces | Lower, more efficient for large spaces     |
| **Determinism**        | Deterministic results                      | Non-deterministic unless a random seed is set |
| **Use Case**           | Suitable for smaller, well-defined hyperparameter spaces | Ideal for larger spaces or limited resources |
| **Execution Time**     | Longer due to exhaustive combinations      | Shorter, as it samples a fixed number of combinations |

